public/app/features/query/state/QueryRunner.ts

```diff
@@ -0,0 +1,115 @@
+import {
+  CoreApp,
+  DataQueryRequest,
+  DataSourceApi,
+  PanelData,
+  rangeUtil,
+  ScopedVars,
+  QueryRunnerOptions,
+  QueryRunner as QueryRunnerSrv,
+} from '@grafana/data';
+import { getTemplateSrv } from '@grafana/runtime';
+import { getDatasourceSrv } from 'app/features/plugins/datasource_srv';
+import { cloneDeep } from 'lodash';
+import { Observable, ReplaySubject, Unsubscribable } from 'rxjs';
+import { getNextRequestId } from './PanelQueryRunner';
+import { preProcessPanelData, runRequest } from './runRequest';
+export class QueryRunner implements QueryRunnerSrv {
+  private subject: ReplaySubject<PanelData>;
+  private subscription?: Unsubscribable;
+  private lastResult?: PanelData;
+
+  constructor() {
+    this.subject = new ReplaySubject(1);
+  }
+
+  get(): Observable<PanelData> {
+    return this.subject;
+  }
+
+  async run(options: QueryRunnerOptions): Promise<void> {
+    const {
+      queries,
+      timezone,
+      datasource,
+      panelId,
+      app,
+      dashboardId,
+      timeRange,
+      timeInfo,
+      cacheTimeout,
+      maxDataPoints,
+      scopedVars,
+      minInterval,
+    } = options;
+
+    if (this.subscription) {
+      this.subscription.unsubscribe();
+    }
+
+    const request: DataQueryRequest = {
+      app: app ?? CoreApp.Unknown,
+      requestId: getNextRequestId(),
+      timezone,
+      panelId,
+      dashboardId,
+      range: timeRange,
+      timeInfo,
+      interval: '',
+      intervalMs: 0,
+      targets: cloneDeep(queries),
+      maxDataPoints: maxDataPoints,
+      scopedVars: scopedVars || {},
+      cacheTimeout,
+      startTime: Date.now(),
+    };
+
+    // Add deprecated property
+    (request as any).rangeRaw = timeRange.raw;
+
+    try {
+      const ds = await getDataSource(datasource, request.scopedVars);
```

Comment:
If we transform this into an Observable, then this function didn't need to be async. But maybe that is taking it a step too far.

Reply:
hmm, what do you think @torkelo would it be better to expose an Observable instead of a Promise for the `QueryRunner.run`?

Reply:
No need to expose Observable

Reply:
I think it then should be the same observable as returned by get() function (ie the subject)

Reply:
The function would just return void in my mind, but if this is due to beta2 then we should skip this.

Reply:
Yeah, I misunderstood @hugohaggmark. I agree with the void scenario.